File: backend/internal/constants/plan.go
package constants

type PlanStatus string

const (
	PlanDraft            PlanStatus = "draft"
	PlanModeled          PlanStatus = "modeled"
	PlanPendingReview    PlanStatus = "pending_supervisor_review"
	PlanApprovedTraining PlanStatus = "approved_for_training"
	PlanArchived         PlanStatus = "archived"
)

var planTransitions = map[PlanStatus]map[PlanStatus]bool{
	PlanDraft:            {PlanModeled: true},
	PlanModeled:          {PlanDraft: true, PlanPendingReview: true},
	PlanPendingReview:    {PlanDraft: true, PlanApprovedTraining: true},
	PlanApprovedTraining: {PlanArchived: true},
	PlanArchived:         {},
}

func ValidPlanStatus(status PlanStatus) bool {
	_, ok := planTransitions[status]
	return ok
}

func CanTransitionPlan(from, to PlanStatus) bool {
	return planTransitions[from][to]
}

func PlanStatuses() []PlanStatus {
	return []PlanStatus{PlanDraft, PlanModeled, PlanPendingReview, PlanApprovedTraining, PlanArchived}
}

File: frontend/src/types/plan.ts
export type PlanStatus =
  | 'draft'
  | 'modeled'
  | 'pending_supervisor_review'
  | 'approved_for_training'
  | 'archived'

export const PLAN_STATUSES: PlanStatus[] = [
  'draft',
  'modeled',
  'pending_supervisor_review',
  'approved_for_training',
  'archived',
]

export interface GasMix {
  o2: number
  he: number
  n2: number
}

export interface DivePlan {
  id: number
  plan_code: string
  diver_profile_id: number
  diver_profile_code?: string
  worksite_pressure_bar: number
  breathing_mix: GasMix
  plan_status: PlanStatus
  created_by: number
  reviewed_by?: number
  version: number
  planned_at: string
  created_at: string
  updated_at: string
}

export interface CreateDivePlan {
  plan_code: string
  diver_profile_id: number
  worksite_pressure_bar: number
  breathing_mix: GasMix
  planned_at: string
}
